Improve netlist connection algorithm

tvt173 opened the pull request:
Hi @joamatab , I rewrote the `get_netlist()` to be more robust and to warn about more issues in optical routing.

**Key changes and improvements:**
- we now do _two_ sweeps over the connections... first we try to connect everything assuming perfect connections at each port. Then we gather ports which did _not_ perfectly connect to anything and try to find imperfect connections, by grouping ports on a coarse grid. This approach does a much better job with such imperfect connections from my trials (and added tests)
- I added a bunch of tests, which should illustrate the new behavior (you can run them against the old algorithm also to see the differences)
- Netlisting with ComponentArray now behaves as you would expect... i.e. an array of 5 straights will show that there are 5 straights in the circuit at the lowest level
- warnings collected during netlisting are reported back into the netlist. These include warnings about mismatched port widths, orientations, shear angles, excessive offsets, etc. You can also configure warning types which should throw an error when encountered by modifying `DEFAULT_CRITICAL_CONNECTION_ERROR_TYPES`. Validators, which will produce warnings for each port type, can be overridden with `DEFAULT_CONNECTION_VALIDATORS`
- A key difference in this algorithm is that we group each port type independently. This allows us to use different logic to determine i.e. if an electrical port is properly connected vs an optical port. In this revision, the core logic is the same, but we employ extra validation for optical ports.
- `snap_to_grid()` now allows a value of 0, which will return the original value, is more efficient when the value is 1, and will throw a more descriptive error when the value is <0
- the signature to `get_netlist()` is exactly the same as it was before, and this should be a drop-in-replacement, however I increased the default value of `tolerance` to 5, because it should allow better performance with the two-grid-sweep approach, as described

joamatab commented:
Hi Troy 

Looks great 

Thank you

Can you make sure the tests pass?

I can merge and release tomorrow

tvt173 commented:
hi @joamatab, thanks. i fixed the tests, and i also pushed a patch to your conftest.py that makes pytest-regressions run significantly faster on windows 🙂 . note that the only regressions are added warnings to some netlists. there was one test you had that was previously passing but now fails because the widths of the connected waveguides did not match. i replaced with one test that asserts that an error is thrown for that case and another which uses waveguides of the same width. i also had to add "warnings" as an allowable top-level attribute of a netlist
